src/Swarm/Language/Typecheck.hs

```diff
@@ -508,3 +541,195 @@ check t ty = do
+  SPair s1 s2 -> (+) <$> analyzeAtomic locals s1 <*> analyzeAtomic locals s2
+  -- Special case for if: number of tangible commands is *max* of branches
+  -- instead of sum, since exactly one of them will be executed
+  SApp (STerm (SApp (STerm (SApp (STerm (TConst If)) tst)) thn)) els ->
+    (+) <$> analyzeAtomic locals tst <*> (max <$> analyzeAtomic locals thn <*> analyzeAtomic locals els)
+  SApp s1 s2 -> (+) <$> analyzeAtomic locals s1 <*> analyzeAtomic locals s2
+  SBind mx s1 s2 -> (+) <$> analyzeAtomic locals s1 <*> analyzeAtomic (maybe id S.insert mx locals) s2
+  SDelay _ s1 -> analyzeAtomic locals s1
+  -- Variables are allowed if bound locally, or if they have a simple type.
+  TVar x
+    | x `S.member` locals -> return 0
+    | otherwise -> do
+      mxTy <- asks $ Ctx.lookup x
+      case mxTy of
+        -- If the variable is undefined, return 0 to indicate the
+        -- atomic block is valid, because we'd rather have the error
+        -- caught by the real name+type checking.
+        Nothing -> return 0
+        Just xTy -> do
+          -- Use applyBindings to make sure that we apply as much
+          -- information as unification has learned at this point.  In
+          -- theory, continuing to typecheck other terms elsewhere in
+          -- the program could give us further information about xTy,
+          -- so we might have incomplete information at this point.
+          -- However, since variables referenced in an atomic block
+          -- must necessarily have simple types, it's unlikely this
+          -- will really make a difference.  The alternative, more
+          -- "correct" way to do this would be to simply emit some
+          -- constraints at this point saying that xTy must be a
+          -- simple type, and check later that the constraint holds,
+          -- after performing complete type inference.  However, since
+          -- this is much simpler, we'll stick with this until such
+          -- time as we have concrete examples showing that the more
+          -- correct, complex way is necessary.
+          xTy' <- applyBindings xTy
+          if isSimpleUPolytype xTy'
+            then return 0
+            else throwError (InvalidAtomic l (NonSimpleVarType x xTy') t)
+  -- No lambda, `let` or `def` allowed!
+  SLam {} -> throwError (InvalidAtomic l AtomicDupingThing t)
+  SLet {} -> throwError (InvalidAtomic l AtomicDupingThing t)
+  SDef {} -> throwError (InvalidAtomic l AtomicDupingThing t)
+  -- We should never encounter a TRef since they do not show up in
+  -- surface syntax, only as values while evaluating (*after*
+  -- typechecking).
+  TRef {} -> throwError (CantInfer l t)
+
+-- | Constants which are allowed inside an @atomic@ block.  This is
+--   intentionally expressed as a whitelist, even though it's much
+--   longer this way, so that we can't create problems in the future
+--   by adding a constant that is problematic to use with @atomic@ but
+--   forgetting to disallow it.
```

Comment:
I am not sure if there is an ergonomic way to do that, but would it not be better to include this in const info?

This is how I understand it:
- functions take 0 ticks
- intangible commands take 1 tick normally but do not count in atomic
- tangible commands take 1 tick normally and there can be at most one in atomic (if branches are an exception)
  - some of those commands can take a higher number of ticks and should _not_ be in atomic at all

Sounds to me like the safest option would be to add a parameter to `Tangible` that would tell us if the command takes just one tick or can take more. A boolean would work, but maybe a `data Length = Short | Long` would be more descriptive, I am definitely open to other name ideas.

Sorry to nitpick on this so much @byorgey, in case you are tired of this atomic business, we can merge as is and improve the ergonomics later :wink:

Reply:
@xsebek don't apologize, I'm glad for all the nitpicking!  And you're absolutely right, this should be included in the const info, I don't know why that didn't occur to me. :facepalm:   Just to clarify:

* All functions and binary operators are intangible.
* Intangible commands do not take up a tick.  That is, a robot can execute any number of intangible commands in one tick (up to the limit of how many CESK machine steps may be taken per tick).
* Tangible commands take up a tick, that is, a robot can execute at most one of them per tick.
* Since we implement `atomic` by suspending the normal mechanisms for limiting the number of steps per tick, we need to make sure that `atomic` still executes at most one tangible command, and moreover does not take multiple ticks, since otherwise it can cause the game to freeze and/or a robot to do more than it should be able to do in one tick. Hence:
    - An `atomic` block can contain at most one tangible comand
    - Commands that can take multiple ticks cannot be allowed in `atomic` at all

I like the idea of adding `data Length = Short | Long` as an argument to `Tangible`.